aiida/engine/daemon/execmanager.py

```diff
@@ -139,8 +139,12 @@ def upload_calculation(node, transport, calc_info, folder, inputs=None, dry_run=
     for code in input_codes:
         if code.is_local():
             # Note: this will possibly overwrite files
-            for f in code.get_folder_list():
-                transport.put(code.get_abs_path(f), f)
+            for f in code.list_object_names():
```

Comment:
This doesn't work if `list_object_names` contains directories, right? Is that ensured not to happen elsewhere?

Reply:
Yes, it should never contain directotries. If you look at the `Code.set_files` implementation, which is called to store the files of the local code, it will copy all the files to top level without nesting. 
```
    def set_files(self, files):
        """
        Given a list of filenames (or a single filename string),
        add it to the path (all at level zero, i.e. without folders).
        Therefore, be careful for files with the same name!
```
I just noticed that I changed that code when the repository interface changed, and I am not sure it does 100% the same. The new implementation checks `os.path.isfile` which wasn't there explicitly. Old implementation:
```
for f in files:
    self.repository.add_path(f, os.path.split(f)[1])
```
However that also does not seem compatible with the docstring which says that you can pass both filenames and directories. I don't think that `add_path` was doing the correct hing either.

Reply:
Well, keeping it as files-only seems like a safe thing to do for now -- we can always add directory compatibility if it's needed. Might be good to assert it in the current piece of code, so it's a bit more explicit this is not supported?

Side note: This `Therefore, be careful for files with the same name!` could easily be improved by just checking first there are no duplicate file names, right?

Reply:
Sure, but that seems a job for another time. If ever anyone actually starts even using local codes..